Core/src/V2/RequestHandler.php

```diff
@@ -0,0 +1,132 @@
+<?php
+/**
+ * Copyright 2023 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Core\V2;
+
+use Google\ApiCore\Serializer;
+use Google\Cloud\Core\EmulatorTrait;
+use Google\Cloud\Core\GrpcRequestWrapper;
+use Google\Cloud\Core\GrpcTrait;
+
+class RequestHandler
+{
+    use EmulatorTrait;
+    use GrpcTrait;
+
+    /**
+     * @var array
+     */
+    private $clientConfig;
+
+    /**
+     * @var array
+     */
+    private $gapics;
+
+    private $serializer;
+
+    /**
+     * @param array $config
+     */
+    public function __construct(
+        Serializer $serializer,
+        array $gapicClasses = [],
+        array $config = []
+    ) {
+        //@codeCoverageIgnoreStart
+
+        $config['serializer'] = $serializer;
+        $this->serializer = $serializer;
+        $config += ['emulatorHost' => null];
+        // TODO: We should be able to swap out the use of
+        // GrpcRequestWrapper with either something in gax, or
+        // have the functionality in this file itself.
+        $this->setRequestWrapper(new GrpcRequestWrapper($config));
+        $grpcConfig = $this->getGaxConfig(
+            $this->pluck('libVersion', $config),
+            isset($config['authHttpHandler'])
+                ? $config['authHttpHandler']
+                : null,
+            $config['transport'] ?? $this->getDefaultTransport()
+        );
+
+        if (isset($config['apiEndpoint'])) {
+            $grpcConfig['apiEndpoint'] = $config['apiEndpoint'];
+        }
+
+        if ((bool) $config['emulatorHost']) {
+            $grpcConfig = array_merge(
+                $grpcConfig,
+                $this->emulatorGapicConfig($config['emulatorHost'])
+            );
+        }
+        //@codeCoverageIgnoreEnd
+
+        $this->clientConfig = $grpcConfig;
+
+        // initialize the gapics
+        foreach($gapicClasses as $gapicClass) {
+            $this->gapics[$gapicClass] = new $gapicClass($this->clientConfig);
```

Comment:
There is client configuration which cannot be shared between clients (e.g. `endpoint`). For that reason, we can't really combine the configuration for the creation of these clients like is being done here.

Reply:
So, the client configurations are shared between the GAPICs even now. We pass the config to the `PubSubClient` which forwards it to `Grpc`. Inside Grpc the same config is used to instantiate the GAPIC classes. So, the `apiEndpoint` if passed as a config, will be shared b/w the GAPICs even in the current state.

I have missed one thing, which I will correct, i.e. if the user passes their own Gapic clients, we need to use them here. I'll take care of that.

Reply:
Since the initialization of these GAPICs has now been shifted to either the PubSubClient or Topic/Subscription, I think this issue is resolved now. The same ClientConfig is passed from PubSubClient to Topic, so there shouldn't be any unintended consequences.